RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByCandidateRegion.cc

```diff
@@ -0,0 +1,304 @@
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/global/EDProducer.h"
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+
+#include "DataFormats/TrackReco/interface/TrackFwd.h"
+#include "DataFormats/TrackReco/interface/Track.h"
+#include "DataFormats/VertexReco/interface/VertexFwd.h"
+#include "DataFormats/VertexReco/interface/Vertex.h"
+#include "DataFormats/BeamSpot/interface/BeamSpot.h"
+#include "DataFormats/Candidate/interface/Candidate.h"
+
+#include "RecoTracker/TkMSParametrization/interface/PixelRecoUtilities.h"
+
+#include <vector>
+#include <memory>
+
+namespace {
+  class TrackSelectorByCandidateRegion final : public edm::global::EDProducer<> {
+  public:
+    typedef enum { BEAM_SPOT_FIXED, BEAM_SPOT_SIGMA, VERTICES_FIXED, VERTICES_SIGMA } Mode;
+
+    explicit TrackSelectorByCandidateRegion(const edm::ParameterSet& conf)
+        : tracksToken_(consumes<reco::TrackCollection>(conf.getParameter<edm::InputTag>("tracks"))),
+          beamspotToken_(consumes<reco::BeamSpot>(conf.getParameter<edm::InputTag>("beamspot"))) {
+      edm::ParameterSet trackPSet = conf.getParameter<edm::ParameterSet>("TrackPSet");
+      minPt_ = trackPSet.getParameter<double>("minPt");
+      edm::ParameterSet chi2_vs_pt = trackPSet.getParameter<edm::ParameterSet>("chi2vsPt");
+      bool enabled = chi2_vs_pt.getParameter<bool>("enabled");
+      pt1_ = enabled ? chi2_vs_pt.getParameter<double>("pt1") : -1.;
+      pt2_ = enabled ? chi2_vs_pt.getParameter<double>("pt2") : -1.;
+      value1_ = enabled ? chi2_vs_pt.getParameter<double>("value1") : -1.;
+      value2_ = enabled ? chi2_vs_pt.getParameter<double>("value2") : -1.;
+
+      if (enabled && pt1_ >= pt2_)
+        throw cms::Exception("Configuration") << "TrackSelectorByCandidateRegion::chi2vsPt: pt1 (" << pt1_
+                                              << ") needs to be smaller than pt2 (" << pt2_ << ")";
+      if (enabled && pt1_ <= 0)
+        throw cms::Exception("Configuration")
+            << "TrackSelectorByCandidateRegion::chi2vsPt: pt1 needs to be > 0; is " << pt1_;
+      if (enabled && pt2_ <= 0)
+        throw cms::Exception("Configuration")
+            << "TrackSelectorByCandidateRegion::chi2vsPt: pt2 needs to be > 0; is " << pt2_;
+
+      //      for (auto const & ir : conf.getParameter<edm::VParameterSet>("regions")) {
+      edm::ParameterSet regionPSet = conf.getParameter<edm::ParameterSet>("RegionPSet");
```

Comment:
It looks like this part here and the implementation should be simply calling CandidateSeededTrackingRegionsProducer (via ES or directly).
This will eliminate the reimplementation of already existing code.